### PR TITLE
Move validation into create mixin

### PR DESCRIPTION
**Pull request by neb42**
Jira ticket: JIRA-CORE24-#

### CODE REVIEW

File: libs/clients/src/lib/prm/prm-base.client.ts [hunk 1]
```diff
@@ -32,14 +33,26 @@ export const BasePrmClient = (entityType: EntityType) =>
...
+  ) =>
   <TBase extends GConstructor<TBasePrmClient>>(Base: TBase) => {
     return class extends Base {
+      public definitionSchema: z.ZodType<TEntityDefinition> = definitionSchema;
+
       public async create(entity: TEntityDefinition): Promise<TEntity> {
         const entitySchema = getEntitySchema(this.entityType);
         const response = await this.post(`/prm/${this.entityType}`, entity);
         return entitySchema.parse(response.data) as TEntity;
       }
```
Comment (neb42) [inline libs/clients/src/lib/prm/prm-base.client.ts:47]: The `create` method is still available for when we want to use the specific service directly

File: libs/clients/src/lib/prm/prm-base.client.ts [hunk 2]
```diff
@@ -32,14 +33,26 @@ export const BasePrmClient = (entityType: EntityType) =>
...
+
+      public validateDefinition(entityDefinition: unknown): TEntityDefinition {
+        return this.definitionSchema.parse(entityDefinition);
+      }
+
+      public validateAndCreate(entityDefinition: unknown): Promise<TEntity> {
```
Comment (neb42) [inline libs/clients/src/lib/prm/prm-base.client.ts:53]: Taking `unknown` as an argument allows it to be the same across all services, making the controller/hook work.

File: libs/clients/src/lib/prm/prm-base.client.ts [hunk 3]
```diff
@@ -32,14 +33,26 @@ export const BasePrmClient = (entityType: EntityType) =>
...
+        return this.create(this.validateDefinition(entityDefinition));
```
Comment (neb42) [inline libs/clients/src/lib/prm/prm-base.client.ts:54]: Passing a known validation function into `create` ensures that it receives the correct type

File: libs/prm-components/src/lib/config/index.ts [hunk 4]
```diff
@@ -19,6 +19,7 @@ const emptyConfig = {
 
 export const configLoader: PrmUIConfigLoader = (staticData) => ({
   [EntityType.Individual]: individualConfig(staticData),
+  [EntityType.Household]: emptyConfig,
```
Comment (neb42) [inline libs/prm-components/src/lib/config/index.ts:22]: Just adding this so the build would build and I could check things

File: libs/prm-components/src/lib/hooks/useCreateEntity.hook.ts [hunk 5]
```diff
@@ -34,7 +34,7 @@ export const useCreateEntity = (
     }
     try {
       actions.submitting();
-      const entity = await client.create(entityDefinition);
+      const entity = await client.validateAndCreate(entityDefinition);
```
Comment (neb42) [inline libs/prm-components/src/lib/hooks/useCreateEntity.hook.ts:37]: We don't really need to validate here as the form should have already validated it. But this is the only way I could make it work.

**Comment by neb42** [comment]
Similar things will be needed for update. List, read, delete should be fine (I think)

File: libs/clients/src/lib/prm/prm-base.client.ts [hunk 6]
```diff
@@ -32,14 +33,26 @@ export const BasePrmClient = (entityType: EntityType) =>
...
     return class extends Base {
+      public definitionSchema: z.ZodType<TEntityDefinition> = definitionSchema;
+
       public async create(entity: TEntityDefinition): Promise<TEntity> {
         const entitySchema = getEntitySchema(this.entityType);
         const response = await this.post(`/prm/${this.entityType}`, entity);
         return entitySchema.parse(response.data) as TEntity;
       }
+
+      public validateDefinition(entityDefinition: unknown): TEntityDefinition {
```
Comment (internetti) [inline libs/clients/src/lib/prm/prm-base.client.ts:49]: why is this its own function? do you expect us to need that somewhere else? but then we could just use parse directly, couldn't we?

**Review by dfigueroa87 — APPROVED** [review]
I'm fine with this approach. Not great from the design perspective as we already discussed, but still good enough.

The only alternatives I could think of are:
- In the controller type the validated definition as `any`, which is definitely not great, but I think still harmless, as we do know what's happening there and would be easy to spot issues.
- Move a bit away from the generalisation, and define specific functions for each entity where needed. We could still have most common logic once (by either inheritance or composition), but would introduce a bit of extra code. The plus side is simplification.

I think for now we can move forward with these changes, and whenever we hit another wall or things continue getting more complex we can evaluate these or other approaches.

**Pull request merged**